Create Blog Post “evolving-the-vonage-helpdesk-with-laravel-deepgram” (#2468)
Nexmo/deved-platform: content/blog/en/evolving-the-vonage-helpdesk-with-laravel-deepgram.md

---

> title: Evolving the Vonage Laravel Helpdesk with OpenAI

Reviewer: Question, I see deepgram mentioned in the blog post md file and images but no deepgram is mentioned in the blog post. Am I missing something here? Should we add anything about Deepgram?

Reviewer 2: This might be a bit tricky to fix. the MD file is named after the original title, which has changed to OpenAI
I think just changing the title might break the build though

Reviewer: @DanyF-github  is aware, and she will fix it at publishing time.

description: In part one of this series, we introduced Helpdesk with SMS. Now,

>   we move onto Voice

Reviewer: ```suggestion
  we move on to Voice
```

thumbnail: /content/blog/evolving-the-vonage-helpdesk-with-laravel-deepgram/laravel-vonage-helpdesk.png
author: james-seconde
published: true
published_at: 2023-05-30T09:58:49.801Z
updated_at: 2023-05-30T09:58:49.831Z
category: tutorial
tags:
  - laravel
  - voice-api
  - open-ai
comments: true
spotlight: false
redirect: ""
canonical: ""
outdated: false
replacement_url: ""
---
This is the second part of a continuous series about an evolving [Laravel](https://laravel.com/) application that leverages Vonage APIs to replicate common real-world use cases.

> In the first part, we created a new Laravel application, pulled in the [vonage-laravel](https://github.com/Vonage/vonage-laravel) library, and created a Helpdesk ticket view. The customer selects their chosen communication method (in this case, it was SMS only) so that any message posted by an admin would be sent to the customer's mobile phone. Replying to the message would then be added using incoming webhooks and written to the ticket conversation.

Reviewer: Would be nice to have the link to the first part here; as I was reading, I was annoyed and looking for the link for the first part. If I have to look for it too much, you'd lose my reading view. Please make my lazy life easy! What do you think?

Reviewer 2: I actually meant to add it, which is why I worded it this way! I'll get the links in

In this article, we're going to add the capability of using Vonage's Text-to-speech (TTS) Voice capabilities using the [Voice API](https://developer.vonage.com/en/voice/voice-api/overview), with the ability of the customer to speak a response that gets transcribed back to the ticket conversation.

### Prerequisites

We'll assume that the first tutorial has been completed, which will give us:

> * The [helpdesk](https://github.com/Vonage-Community/sample-messages_voice-php-helpdesk) repository, cloned locally from Github

Reviewer: ```suggestion
* The [helpdesk](https://github.com/Vonage-Community/sample-messages_voice-php-helpdesk) repository, cloned locally from GitHub
```

* [Laravel Sail](https://laravel.com/docs/10.x/sail) up and running, to dockerise the local development environment
* [Migrations run](https://laravel.com/docs/10.x/migrations)
* [Vite](https://vitejs.dev/) development server running to build [Laravel Breeze](https://laravel.com/docs/10.x/starter-kits)'s boilerplate assets
* [Ngrok](https://ngrok.com/) running locally, and a Vonage application instance configured to send webhooks to it

> ### How does it do that? Part 2: Voice

Reviewer: As per our guidelines https://learn.vonage.com/contributing/ we follow this title styling.
```suggestion
### How Does It Do That? Part 2: Voice
```

> OK, so it's time to go through Voice capabilities. The flow of how conversations happen here is exactly the same: you create a new ticket as a customer and it opens the conversation view. However, when we create it this time, we're going to set up the ticket as a voice conversation. 

Reviewer: Nice to have: As I was reading this, I felt that if there was an image of what you're saying, it could help me visualize it.

### Enabling Voice

> Before this will work, we're going to need a Voice-enabled application ID in the [Vonage Dashboard](https://dashboard.nexmo.com/). You can either edit the previous application from the last tutorial or create a new one. Enabling the application ID  is all you need: don't worry about using the UI to send webhooks to the correct local route (we'll go through this later) as the code constructs the webhook reply URL for you (this is different from how we've set up SMS, and I'll go through why later).

Reviewer: ```suggestion
Before this works, we will need a Voice-enabled application ID in the [Vonage Dashboard](https://dashboard.nexmo.com/). You can edit the previous application from the last tutorial or create a new one. Enabling the application ID is all you need: don't worry about using the UI to send webhooks to the correct local route (we'll go through this later) as the code constructs the webhook reply URL for you (this is different from how we've set up SMS, and I'll go through why later).
```

> Before this will work, we're going to need a Voice-enabled application ID in the [Vonage Dashboard](https://dashboard.nexmo.com/). You can either edit the previous application from the last tutorial or create a new one. Enabling the application ID  is all you need: don't worry about using the UI to send webhooks to the correct local route (we'll go through this later) as the code constructs the webhook reply URL for you (this is different from how we've set up SMS, and I'll go through why later).

Reviewer: Later when? Maybe specify in which session? Don't do this to my anxious curious heart!

Reviewer 2: I've reworded this a bit

### What is OpenAI?

The most common product you may have heard of in association with this name is [ChatGPT](https://chat.openai.com/).

ChatGPT is a cutting-edge language model developed by [OpenAI](https://openai.com/), designed to engage in natural language conversations with users. As an AI-based assistant, ChatGPT can provide information, answer questions, and assist with various tasks. It uses deep learning techniques to understand and generate human-like text, making the interactions feel more personalized.

> Yes, that paragraph was written by ChatGPT. But what you might not know is the company behind it, OpenAI, have [several other products](https://openai.com/product) that can all be accessed via. their API. One such product is [Whisper](https://platform.openai.com/docs/guides/speech-to-text), which we're going to use to transcribe what a customer says as a recorded message in response to a ticket entry from the Helpdesk app.

Reviewer: ```suggestion
Yes, that paragraph was written by ChatGPT. But what you might not know is the company behind it, OpenAI, has [several other products](https://openai.com/product) that can all be accessed via. their API. One such product is [Whisper](https://platform.openai.com/docs/guides/speech-to-text), which we're going to use to transcribe what a customer says as a recorded message in response to a ticket entry from the Helpdesk app.
```

> Yes, that paragraph was written by ChatGPT. But what you might not know is the company behind it, OpenAI, have [several other products](https://openai.com/product) that can all be accessed via. their API. One such product is [Whisper](https://platform.openai.com/docs/guides/speech-to-text), which we're going to use to transcribe what a customer says as a recorded message in response to a ticket entry from the Helpdesk app.

Reviewer: Are you sure this is the correct link to Whisper?

Reviewer 2: somewhat confusingly, yes it is (the model is called Whisper, but it's also referred to as speech-to-text)

### Setting up the OpenAI API

> Firstly, you'll need an OpenAI account. Follow this link to create an account, and when you're done you'll need to head to 'Manage Account' under your top right hand profile menu. Opening this up will show you the following screen - head to API Keys and set up a new key. The end result should look something like this:

Reviewer: WHICH LINK? AHHH! haha please add the link.

> ![](/content/blog/evolving-the-vonage-laravel-helpdesk-with-openai/screenshot-2023-06-02-at-10.22.45.png)

Reviewer: Please add an Alt image description.

When you create the key, you'll have one chance to copy it - make sure you do that.

We need to add that secret to our `env` file. You can see in the `example.env` file in the repo that we have a placeholder for it:

> ![](/content/blog/evolving-the-vonage-laravel-helpdesk-with-openai/screenshot-2023-06-02-at-10.43.28.png)

Reviewer: Please add an Alt image description.

Reviewer: There is 'Deepgram' in the image but no description in the alt image or anywhere in the blog post about Deepgram.

Reviewer 2: This used to be an article using Deepgram, but I swapped it for OpenAI

I've included the others in the screenshot because it's important to note that this feature won't work without all of these environment variables set:

* `VONAGE_SMS_FROM` is re-used as the outbound calling number
* `PUBLIC_URL` is your [Ngrok](https://ngrok.com/) (or any other tool such as Beyond Code's [Expose](https://expose.dev/docs/introduction)) public-facing address. This is essential, as the code will stitch together the response URL to the API when making a call

> * `VONAGE_APPICATION_ID` and `VONAGE_PRIVATE_KEY`. In the last tutorial, we could have used basic authentication, but for webhooks to work they need to be tied to an application ID. For using the Vonage Voice API, we have to have a private key and application ID, which the Vonage PHP SDK will use to generate and handle JWT authorisation for us.

Reviewer: ```suggestion
* `VONAGE_APPICATION_ID` and `VONAGE_PRIVATE_KEY`. In the last tutorial, we could have used basic authentication, but for webhooks to work, they need to be tied to an application ID. For using the Vonage Voice API, we have to have a private key and application ID, which the Vonage PHP SDK will use to generate and handle JWT authorisation for us.
```

* This is the neat bit. You know how in part one of this tutorial, we set an Ngrok URL in the Vonage Dashboard for SMS webhooks? We've not done that here, because each call using the Voice SDK can be configured to *use a specific callback URL for this call we're making*. This part is really important because it *allows us to configure state.* In this case, we take the Ngrok public URL from `$currentHost` (i.e. the constant `PUBLIC_URL`), a route defined by us for our application (`/webhook/answer/`) and the key to making this work: the ticket entry ID as part of the route. Later, in the `WebhookController` we can pull the parent ticket out, plus the owner of that ticket.

So, now we need a new controller to handle what comes in when the customer has completed their ticket call. The two parts to this are:

* Read out a response when the customer picks up their phone call (this will be set in the controller assigned to the route)
* Have a route to read incoming answer events (we set these when setting up the outbound call)

> * From the recording event generated after the call is complete, fetch a voice recording of the customers' response, transcribe it using OpenAI, and write it as a new `TicketEntry`.

Reviewer: This item on the list ends with a period, but the others don't.

The route points to `WebhookController::answer()`, so our TTS response looks like this:

```php
public function answer(TicketEntry $ticketEntry): JsonResponse  
{  
    if (!$ticketEntry->exists) {  
        return response()->json([  
            [                'action' => 'talk',  
                'text' => 'Sorry, there has been an error fetching your ticket information'  
            ]  
        ]);    }  
    return response()->json([  
        [            'action' => 'talk',  
            'text' => 'This is a message from the Vonage Helpdesk'  
        ],  
        [            'action' => 'talk',  
            'text' => $ticketEntry->content,  
        ],        [            'action' => 'talk',  
            'text' => 'To add a reply, please leave a message after the beep, then press the pound key',  
        ],        [            'action'    => 'record',  
            'endOnKey'  => '#',  
            'beepStart' => true,  
            'eventUrl' => [config('helpdesk.public_url') . '/webhook/recordings/' .  $ticketEntry->id]  
        ],        [            'action' => 'talk',  
            'text' => 'Thank you, your ticket has been updated.',  
        ]    ]);}
```

> Each array gives a payload of instructions that are fairly straightforward, but the important glue here to answer the question "how do we capture the customer's response?" is in the `record` action. You can see it gives a `beepStart` prompt, and most importantly we define the behaviour after the call is complete. The `eventUrl` will be hit with a webhook that will contain a URL of this recording.

Reviewer: ```suggestion
Each array gives a payload of instructions that are fairly straightforward, but the important glue here to answer the question "How do we capture the customer's response?" is in the `record` action. You can see it gives a `beepStart` prompt, and most importantly we define the behaviour after the call is complete. The `eventUrl` will be hit with a webhook that will contain a URL of this recording.
```

### Processing the Recording

> Our next route is the one that will contain a link to the customer's response as a recorded MP3, as well as the ticket ID so we know what entity it belongs to. Here's an example payload we can expect to receive:

Reviewer: ```suggestion
Our next route is the one that will contain a link to the customer's response as a recorded MP3, as well as the ticket ID, so we know what entity it belongs to. Here's an example payload we can expect to receive:
```

  "end_time": "2020-01-01T12:00:00.000Z",
  "conversation_uuid": "CON-aaaaaaaa-bbbb-cccc-dddd-0123456789ab",
  "timestamp": "2020-01-01T12:00:00.000Z"
}
```

And here is our route:

```php
Route::post('/webhook/recordings/{ticketEntry:id}', [WebhookController::class, 'recording'])->name('voice.recording');
```

And the `recording()` method to handle it:

```php
public function recording(TicketEntry $ticketEntry, Request $request): Response|Application|ResponseFactory  
{  
    $params = $request->all();  
    Log::info('Recording event', $params);  
  
    $audio = Vonage::voice()->getRecording($params['recording_url']);  
    Storage::put('call_recording.mp3', $audio);  
  
    $ticketContent = $this->transcribeRecordingOpenAi();  
  
    $newTicketEntry = new TicketEntry([  
        'content' => $ticketContent,  
        'channel' => 'voice',  
    ]);  
    $parentTicket = $ticketEntry->ticket()->get()->first();  
    $newTicketEntryUser = $parentTicket->user()->get()->first();  
    $newTicketEntry->user()->associate($newTicketEntryUser);  
    $newTicketEntry->ticket()->associate($parentTicket);  
    $newTicketEntry->save();  
  

>     return response('', 204);  

Reviewer: At the end of all the lines of the code, there are double spaces. Not sure it's a PHP thing, but it seems wrong to me.

Reviewer 2: this is a formatting error for me to fix

}
```

This uses Route Model Binding to take the relevant `TicketEntry` and inject it as a dependency, then pulls out the `recording_url`. The Vonage SDK has a useful method named `getRecording()` that will return a `StreamInterface` contained in the body. 

> For security reasons, you cannot write out the stream from the audio straight to the OpenAI request we're going to send for transcription, so we need to save the file temporariy. Once we've saved it, we can then use the `Storage` facade to read it back out during the transcription request and then delete it.

Reviewer: ```suggestion
For security reasons, you cannot write out the stream from the audio straight to the OpenAI request we will send for transcription, so we need to save the file temporarily. Once we've saved it, we can use the `Storage` facade to read it back out during the transcription request and then delete it.
```

> The `transcribeRecording()` is a custom method in this class controller that we'll get to in a moment, but assuming that a string comes back from the transcription, we create a new `TicketEntry`, associate it with the ticket owner (we know this is the customer as it's an incoming webhook route), and save it to the `Ticket`

Reviewer: ```suggestion
The `transcribeRecording()` is a custom method in this class controller that we'll get to in a moment, but assuming that a string comes back from the transcription, we create a new `TicketEntry`, associate it with the ticket owner (we know this is the customer as it's an incoming webhook route), and save it to the `Ticket`.
```

        ],        'body' => $multipart,  
    ]);  
    Storage::delete('call_recording.mp3');  
  
    $responseBody = json_decode($response->getBody()->getContents(), true, 512, JSON_THROW_ON_ERROR);  
  
    return $responseBody['text'];  
}
```

This has been hacked together for demonstration reasons, so firstly I'd like to make it clear that if your application has a dependency on a 3rd party API such as this one (or Vonage), you should wrap this client and its configuration [as a service provider](https://laravel.com/docs/10.x/providers).

The method creates a new [Guzzle](https://docs.guzzlephp.org/en/stable/) Client, and prepares the request as a `MultipartStream`, as OpenAI requires the request to be an encoded form. We set the base url and fetch our temporary file we created earlier (`call_recording.mp3`). We can now use `fopen()` to write the file out, and then delete it after the request has been completed. 

All being well, you'll get a transcription array back, which will contain the key `text`, which is sent back for updating the `TicketEntry`. Congratulations: we now have a working TTS ticketing system!

### And, Stay Tuned

> There are two more tutorials forthcoming, so keep an eye out for them. None of our code is tested, so [PEST](https://pestphp.com/) will come to the rescue for that. We also are using plain [Blade](https://laravel.com/docs/10.x/blade) templates - what if we want to instantly update the page as if it was a JavaScript Single-Page-Application (SPA)? Well, we have [Livewire](https://laravel-livewire.com/) for that. Until the next time, enjoy helping your customers!

Reviewer: Please add a note inviting folks to our Twitter and Slack community as well.

https://vonage.slack.com/archives/C0200MQ4BPE/p1655900662081889

Reviewer: Please don't forget to add this last part!

ChatGPT is a cutting-edge language model developed by [OpenAI](https://openai.com/), designed to engage in natural language conversations with users. As an AI-based assistant, ChatGPT can provide information, answer questions, and assist with various tasks. It uses deep learning techniques to understand and generate human-like text, making the interactions feel more personalized.

Yes, that paragraph was written by ChatGPT. But what you might not know is the company behind it, OpenAI, has [several other products](https://openai.com/product) that can all be accessed via. their API. One such product is [Whisper](https://platform.openai.com/docs/guides/speech-to-text), which we're going to use to transcribe what a customer says as a recorded message in response to a ticket entry from the Helpdesk app.

### Setting up the OpenAI API

Firstly, you'll need an OpenAI account. [Follow this link to create an account](https://auth0.openai.com/u/signup), and when you're done you'll need to head to 'Manage Account' under your top right hand profile menu. Opening this up will show you the following screen - head to API Keys and set up a new key. The end result should look something like this:

> ![Screenshot of the OpenAI dashboard showing API Key creation](/content/blog/evolving-the-vonage-laravel-helpdesk-with-openai/screenshot-2023-06-02-at-10.22.45.png)

Reviewer: Hello! One more point about images. They cannot be named 'screenshot' etc. They have to have meaningful names. @SecondeJK

When you create the key, you'll have one chance to copy it - make sure you do that.

We need to add that secret to our `env` file. You can see in the `example.env` file in the repo that we have a placeholder for it:

> ![Screenshot of Helpdesk's environment variables example file](/content/blog/evolving-the-vonage-laravel-helpdesk-with-openai/screenshot-2023-06-02-at-10.43.28.png)

Reviewer: I see you've updated the correct file! Could you please remove the file 'B 
[...lving-the-vonage-helpdesk-with-laravel-deepgram/pasted-image-20230526211505.png](https://github.com/Nexmo/deved-platform/pull/2468/files#diff-deee2062646e46c21cb19c5aff85a1a595f029c4d60ef761b9a23c0068adc982)' (the one that contains the deepgram image, from this PR?

> ![Screenshot of Helpdesk's environment variables example file](/content/blog/evolving-the-vonage-laravel-helpdesk-with-openai/screenshot-2023-06-02-at-10.43.28.png)

Reviewer: Hello! One more point about images. They cannot be named 'screenshot' etc. They have to have meaningful names. @SecondeJK
